src/shared/Microsoft.Git.CredentialManager/GitConfiguration.cs

```diff
@@ -180,10 +194,11 @@ public void Enumerate(GitConfigurationLevel level, GitConfigurationEnumerationCa
             }
         }
 
-        public bool TryGet(GitConfigurationLevel level, string name, out string value)
+        public bool TryGet(GitConfigurationLevel level, GitConfigurationType type, string name, out string value)
         {
             string levelArg = GetLevelFilterArg(level);
-            using (Process git = _git.CreateProcess($"config --null {levelArg} {QuoteCmdArg(name)}"))
+            string typeArg = GetCanonicalizeTypeArg(type);
+            using (Process git = _git.CreateProcess($"config --null {levelArg} {typeArg} {QuoteCmdArg(name)}"))
```

Comment:
Which version of Git first supported the `--type=<type>` argument?

I see from the [documentation](https://git-scm.com/docs/git-config#Documentation/git-config.txt---bool) that there used to be (and still is, but deprecated) explicit options for each type instead.

```text
--bool
--int
--bool-or-int
--path
--expiry-date
    Historical options for selecting a type specifier. Prefer instead --type (see above).
```

We don't want to try and use this new argument on an old version of Git or else we'll get usage errors(!) This happened in the past when I added `--show-scope` which does not exist on the (old) version of Git available in some Ubuntu/Debian distributions.

If we can, try and use the oldest still supported arguments. Otherwise we'll need to introduce a `git version` call and check (that could be cached at least).

Reply:
It looks like [2.18](https://git-scm.com/docs/git-config/2.18.0) is the first released version to support the `--type` parameter (although it only supports `--int`, `--bool`, and `--path` in [2.17 and older](https://git-scm.com/docs/git-config/2.17.0) - fine for the purposes of this change, since it's only using the path type anyway). 

Is there a minimum version of git that GCM Core explicitly supports? If not (or if it's older than 2.18), I'm not sure I like the idea of actively using a deprecated option, so an explicit version check would be a nice way to handle this (and helpful for using newer features in git in general, should they ever be needed). I think `Version` (or `GitVersion`?) would make sense as a property of the `Git` class, but I don't see any other properties there - is that an intentional design decision, or has there just not been any need for public properties there yet?

Reply:
Support for `--type=path` (and making it the preferred option) was introduced in git/git@fb0dc3bac135e9f6243bd6d293e8c9293c73b9cd, and the first Git version to include this patch was v2.18.0.

Reply:
The `--path` option was introduced in git/git@1349484e341a3ec2ba02a86c8fbd97ea9dc8c756, the first Git version to support it was v1.7.0. I think that it is sane to require that minimum Git version ;-) (but v2.18.0 might be a bit too new, some Debian LTS versions are still based on v2.7.0 IIRC).

Reply:
Right.. 2.18 is probably too new.

> Is there a minimum version of git that GCM Core explicitly supports?

Historically, no.. not really. On Windows it was "approximately the version that ship together in Git for Windows" but that doesn't hold anymore for Mac/Linux, nor always for Windows too (what with MinGit and VS etc).

A good rule of thumb would be looking at what version is shipped with the previous Ubuntu/Debian LTS.. we don't want to break existing users if possible, and/or what the minimum version of Git is required today. I'm not sure GCM uses _any_ options or commands that were not added a long time ago?

All currently called Git commands and args are:

```text
rev-parse --absolute-git-dir
remote -v show
config (--local|--global|--system) <name> <value>
config (--local|--global|--system) --add <name> <value>
config (--local|--global|--system) --unset <name>
config (--local|--global|--system) --unset-all <name>
config (--local|--global|--system) --replace-all <name> <value> <value-regex>
config --null (--local|--global|--system) <name>
config --null (--local|--global|--system) --list
config --null (--local|--global|--system) --get-all <name>
config --null (--local|--global|--system) --get-regex <name-regex>
```

Adding a version check would future proof us when (or rather _IF_) the old options go away, but I don't think the old options are really going to go away any time soon! @dscho ?

Reply:
They won't be going away any time soon, no. But the `--path` option _is_ deprecated, so I would prefer us to use `--type=path` when possible.

Reply:
The last commit in this series adds a `GitVersion` implementation & selects the right flags in `GitProcessConfiguration` based on the `GitProcess` version. Please let me know what you think of that approach (at the very least to make sure I followed repository conventions properly), or if you'd like me to move it into a separate pull request altogether.

Reply:
Looks good to me!